### PR TITLE
[Fix #8067] Added screens for ENS preferred name selection

### LINKED ISSUE #8067: [ENS Names] Use ens username in chat
Implement ens username settings and show ens names in chats

image

image

crossed element not in the scope of this PR

Acceptance criteria:

1) pixel perfect
2) it should be possible to select a primary username
3) it should be possible to enable "use my ENS name in chats"
4) ens name must be verified and shown in chats for messages which contains ens name

Figma: https://www.figma.com/file/TNCyHKtR3sx5EL6YznFWUa4O/Profile?node-id=510%3A3116

### PR DESCRIPTION
**Pull request by jeluard**
fixes #8067

This PR only contains UI changes (with the exception of ENS not being used in chats)

### Summary

Added screens allowing to defined a preferred ENS name. Persisted

### Steps to test

- Open Status
- Register several ENS names
- Select your favourite one
- Restart; it's still there

### Limitations

- registering a name you own not associated with a your whisper key is not supported
- `Remove username` is not implemented

status: ready

### CODE REVIEW

**Review by vitvly — APPROVED** [review]

File: src/status_im/ui/screens/ens/views.cljs [hunk 1]
```diff
@@ -487,35 +495,77 @@
...
+                         :action-fn     action
+                         :icon          :main-icons/username
+                         :hide-chevron? hide-chevron?}]))
+
+(defn- name-list [names preferred-name bottom-shown?]
+  [react/view {:style {:margin-top 24 :height 320}}
+   [react/view {:style {:margin-horizontal 16 :align-items :center :justify-content :center}}
+    [react/nested-text
+     {:style {:color colors/gray}}
+     "Your messages are displayed to others with"
```
Comment (dmitryn) [inline src/status_im/ui/screens/ens/views.cljs]: shouldn't string be localized?

File: src/status_im/ui/screens/ens/views.cljs [hunk 2]
```diff
@@ -487,35 +495,77 @@
...
+     (i18n/label :t/ens-chat-settings)]
+    (when (seq names)
+      [profile.components/settings-item {:label-kw  :ens-primary-username
+                                         :value     preferred-name
+                                         :action-fn #(reset! bottom-shown? true)}])
+    [profile.components/settings-switch-item {:label-kw  :ens-show-username
+                                              :action-fn #(re-frame/dispatch [:ens/switch-show-username])
+                                              :value     show?}]]
+   (let [message {:username name :from public-key :last-in-group? true :display-username? true :display-photo? true
+                  :content {:text "Hey"} :content-type "text/plain" :timestamp-str "9:41 AM"}]
```
Comment (dmitryn) [inline src/status_im/ui/screens/ens/views.cljs]: what are the "Hey" and "9:41 AM" strings, some default values? need localization?

**Review by dmitryn — APPROVED** [review]

File: src/status_im/ui/screens/ens/views.cljs (hunk 2 above)
Comment (errorists) [reply src/status_im/ui/screens/ens/views.cljs]: Good point. of those two I'd say that 'Hey' that appears in the dummy message should be localised.

**Review by errorists — CHANGES_REQUESTED** [review]
issues I noticed, not all of them are directly related to the scope of this PR but it's still ENS stuff :)

IMG_0986

- when the input is empty, it's missing the description [Figma](https://www.figma.com/file/TNCyHKtR3sx5EL6YznFWUa4O/Profile?node-id=1822%3A1231)

Group1

- in addition to the redlined items, when you have only a single name added, the list item for choosing a preferred username shouldn't be visible [Figma](https://www.figma.com/file/TNCyHKtR3sx5EL6YznFWUa4O/Profile?node-id=1828%3A20296)

IMG_0988

- the bottom sheet is all messed up :D take a look at how it's used in the (+) button for instance, also [Figma](https://www.figma.com/file/TNCyHKtR3sx5EL6YznFWUa4O/Profile?node-id=1828%3A21738)

- [already discussed on status] the first username you add should be defined as the preferred one, so there's no possibility to run into a case where there's no preferred name set.

- toggling the 'Show my ENS username in chats on' doesn't update the message preview, the username should reflect how you will be seen in chat, with ENS, or without. [Figma](https://www.figma.com/file/TNCyHKtR3sx5EL6YznFWUa4O/Profile?node-id=1828%3A20099)

Group 2

- and lastly the Ministry of Silly Margins called, wants its margins here! :)

thanks!

**Comment by jeluard** [comment]
@errorists 

> the bottom sheet is all messed up :D take a look at how it's used in the (+) button for instance, also

What's messed up beyond the height? Tried to fix that BTW but I failed so far. I think we need an expert

**Comment by errorists** [comment]
> @errorists
> 
> > the bottom sheet is all messed up :D take a look at how it's used in the (+) button for instance, also
> 
> What's messed up beyond the height? Tried to fix that BTW but I failed so far. I think we need an expert

It should be rendered as the topmost element covering the top toolbar and bottom tab bar.

**Comment by jeluard** [comment]
Hmm I see now. Thanks!

**Comment by jeluard** [comment]
> toggling the 'Show my ENS username in chats on' doesn't update the message preview, the username should reflect how you will be seen in chat, with ENS, or without

Right this is not implemented yet

**Review by errorists — APPROVED** [review]

**Comment by errorists** [comment]
one last comment, feel free to address in future PRs. 

sheet

**Comment by asemiankevich** [comment]
@jeluard you are probably linked wrong issue to this PR, would you mind to edit?

**Comment by jeluard** [comment]
@asemiankevich Done

**Comment by asemiankevich** [comment]
@jeluard can't perform a transaction to associate the name with chat key

image

**use same acount both develop and PR build**

Steps:
1. open status (develop)
2. tap profile -> ens usernames
3. register any free username
4. on another device (current PR build):  tap Add username and search for name you just registered in develop for same account -> tap blue button to proceed with txn

Expected: screen with terms and conditions appear

**Comment by asemiankevich** [comment]
@jeluard would you mind to resolve conflicts? last builds are failed with your recent commit. Thank you!

**Comment by asemiankevich** [comment]
@jeluard would you mind to make this bottom sheet scrollable so i can access as many names i want (not only first 5?)

image

**Comment by asemiankevich** [comment]
still can't scroll (

image

**Comment by asemiankevich** [comment]
bottom sheet issue and registering names which are not connected to whisper key will be logged separately

**Pull request merged**